dms/models/ir_attachment.py

```diff
@@ -0,0 +1,64 @@
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+from odoo import api, models
+
+
+class IrAttachment(models.Model):
+
+    _inherit = "ir.attachment"
+
+    def _get_dms_directories(self, res_model, res_id):
+        return self.env["dms.directory"].search(
+            [("res_model", "=", res_model), ("res_id", "=", res_id)]
+        )
+
+    def _dms_operations(self, attachment_ids):
+        for attachment in attachment_ids:
+            if attachment.res_model and attachment.res_id:
+                directories = self._get_dms_directories(
+                    attachment.res_model, attachment.res_id
+                )
+                if not directories:
+                    # Auto-create directories
+                    directories_without_res_id = self._get_dms_directories(
+                        attachment.res_model, False
+                    )
+                    for directory in directories_without_res_id:
+                        model_item = (
+                            self.env[attachment.res_model]
+                            .sudo()
+                            .browse(attachment.res_id)
+                        )
+                        self.env["dms.directory"].create(
+                            {
+                                "name": model_item.display_name,
+                                "res_model": attachment.res_model,
+                                "res_id": attachment.res_id,
+                                "parent_id": directory.id,
+                                "storage_id": directory.storage_id.id,
+                            }
+                        )
+                    # Get dms_directories again (with items previously created)
+                    directories = self._get_dms_directories(
+                        attachment.res_model, attachment.res_id
+                    )
+                # Auto-create_files
+                if directories:
+                    for directory in directories:
+                        self.env["dms.file"].create(
+                            {
+                                "name": attachment.name,
+                                "directory_id": directory.id,
+                                "attachment_id": attachment.id,
+                                "res_model": attachment.res_model,
+                                "res_id": attachment.res_id,
+                            }
+                        )
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        if self.env.context.get("dms_file"):
+            return super(IrAttachment, self).create(vals_list)
+        else:
+            res = super(IrAttachment, self).create(vals_list)
+            self._dms_operations(res)
+            return res
```

Comment:
Go to more elegant code:
```suggestion
        records = super().create(vals_list)
        if not self.env.context.get("dms_file"):
            records._dms_operations()
        return records
```